### PR TITLE
Update set_types to also set logical types

### LINKED ISSUE #587: Implement `set_types` on Table Accessor
This allows for updating logical types on the Table Accessor.

As this will cause changes to the underlying DataFrame, we might want to consider returning a new DataFrame and Schema here instead of doing things inplace

### PR DESCRIPTION
**Pull request by tamargrey**
- Updates `df.ww.set_types` so that it can also take in a logical types parameter
- Updates underlying dtype of DataFrame is logical type is changed
- Also removes underlying index if index has been removed from the Schema 
- Closes #587

### CODE REVIEW

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:181]: Felt slightly weird using `_get_column_logical_type` here, since, technically, if `ltype` wasn't defined it'd try to infer the logical type, which we definitely don't want (this is happening in the DataTables' set types as well). To avoid that, I passed `None` in for the series so that we definitely wouldn't quietly infer a LogicalType during `set_types`.

I'd be open to pulling out the logic from this function that parses a logical type user input, but not sure how useful that'd be.

File: woodwork/table_accessor.py
Comment (tamargrey) [inline woodwork/table_accessor.py:196]: Felt it was important to follow our behavior of resetting the underlying index when it's removed, but this isn't getting done for `reset_semantic_tags` or `remove_semantic_tags` since they only exist on the Schema class, and this isn't actually behavior that we want (#597 ), so maybe it's better to not have this here at all.

File: woodwork/tests/schema/test_schema.py
Comment (tamargrey) [inline woodwork/tests/schema/test_schema.py:356]: I don't know why a user would do this, but figured it'd be better to just do nothing than to error or still reset. Happy to discuss other options, though

File: woodwork/schema.py
Comment (tamargrey) [inline woodwork/schema.py:162]: I don't know that we want the `Schema.set_types` to only take in LogicalType objects. It follows the way that the Schema only takes in LogicalType objects, but maybe that's less important here.

File: woodwork/table_accessor.py
Comment (thehomebrewnerd) [reply woodwork/table_accessor.py:181]: My initial reaction is that we should pull out the logic that parses the logical type into a separate function that we could use here and inside `_get_column_logical_type`. That would make the code a bit more understandable I think rather than trying to understand why `None` is being passed for the series here.

File: woodwork/table_accessor.py
Comment (thehomebrewnerd) [reply woodwork/table_accessor.py:196]: TBH the underlying index behavior is getting me a bit confused. I'd vote we have some good discussion about how we really want it to work in all these different scenarios and tackle it separate from this issue.

File: woodwork/tests/accessor/test_table_accessor.py
Comment (thehomebrewnerd) [inline woodwork/tests/accessor/test_table_accessor.py:800]: Maybe move this line outside the context manager so we know clearly which statement we are testing. Not sure why the df creation was inside before...

File: woodwork/tests/accessor/test_table_accessor.py
Comment (thehomebrewnerd) [inline woodwork/tests/accessor/test_table_accessor.py:1360]: Should we also test setting a semantic tag somewhere in this test?

File: woodwork/tests/schema/test_schema_init.py
Comment (thehomebrewnerd) [inline woodwork/tests/schema/test_schema_init.py:95]: Extra blank line

File: woodwork/schema.py
Comment (tamargrey) [reply woodwork/schema.py:162]: Even in the `Schema.__init__`?

I agree that we want to avoid people having to import more LogicalTypes than they have to, but, in theory, they should rarely be directly creating Schemas or calling Schema.set_types, so it probably won't be that big of an issue.

then again, I can't think of a reason that a Schema shouldn't be able to parse logical type strings other than trying to simplify it a bit.

File: woodwork/table_accessor.py
Comment (tamargrey) [reply woodwork/table_accessor.py:181]: Cool. I'm happy to pull that logic out! (and also maybe just move it to the Schema.set_types)

File: woodwork/table_accessor.py
Comment (tamargrey) [reply woodwork/table_accessor.py:196]: I think that's very fair

Comment (tamargrey) [reply woodwork/table_accessor.py:196]: Going to remove this behavior so that there's not inconsistency with index removal via tags

File: woodwork/schema.py
Comment (thehomebrewnerd) [reply woodwork/schema.py:162]: Sorry - I'm getting confused between the methods. I think it is fine for the Schema methods to only take in LogicalType classes, but the accessor methods should be more flexible.

Comment (tamargrey) [reply woodwork/schema.py:162]: Ahh, got it. So to confirm: We'll allow string inputs in `WoodworkTableAccessor.set_types` but not for `Schema.set_types`

Comment (thehomebrewnerd) [reply woodwork/schema.py:162]: Yep, that's what I was thinking.

File: woodwork/tests/accessor/test_table_accessor.py
Comment (tamargrey) [reply woodwork/tests/accessor/test_table_accessor.py:1360]: Makes sense. I'll update the checks to include one that sets semantic tags as well

File: woodwork/tests/schema/test_schema.py
Comment (tamargrey) [reply woodwork/tests/schema/test_schema.py:356]: I think I'd like to catch the None case in `_check_semantic_tags`--that way it'd also get caught at init.

Comment (thehomebrewnerd) [reply woodwork/tests/schema/test_schema.py:356]: Makes sense to me

File: woodwork/schema.py
Comment (tamargrey) [reply woodwork/schema.py:196]: I think we still need some intermediate like `ltype_update_tags` so that `new_semantic_tags` can take precedence when it's present, but if it's `None`, it'll mean that there was an ltype update, so we need `ltype_update_tags` to reset the semantic tags`

Comment (thehomebrewnerd) [reply woodwork/schema.py:195]: Not sure if it is better or not, but this is what I was thinking for the `for` loop:
```python
        for col_name in (logical_types.keys() | semantic_tags.keys()):
            original_tags = self.semantic_tags[col_name]

            # Update Logical Type for the Schema, getting new semantic tags
            new_logical_type = logical_types.get(col_name)
            if new_logical_type is not None:
                self.columns[col_name]['logical_type'] = new_logical_type
                # ltype_update_tags = _reset_semantic_tags(new_logical_type.standard_tags, self.use_standard_tags)

            # Get new semantic tags, removing existing tags
            new_semantic_tags = semantic_tags.get(col_name)
            if new_semantic_tags is not None:
                new_semantic_tags = _set_semantic_tags(new_semantic_tags,
                                                       self.logical_types[col_name].standard_tags,
                                                       self.use_standard_tags)
                _validate_not_setting_index_tags(new_semantic_tags, col_name)
            else:
                new_semantic_tags = _reset_semantic_tags(new_logical_type.standard_tags, self.use_standard_tags)

            self.columns[col_name]['semantic_tags'] = new_semantic_tags

            if retain_index_tags and 'index' in original_tags:
                self._set_index_tags(col_name)
            if retain_index_tags and 'time_index' in original_tags:
                self._set_time_index_tags(col_name)
```

Comment (thehomebrewnerd) [reply woodwork/schema.py:195]: I think that works because we know if `new_semantic_tags` is `None` that means `new_logical_type` is not `None`.

Comment (tamargrey) [reply woodwork/schema.py:195]: Ahh, got it. That'll definitely work, but I think it might be beneficial to keep the actual reset call next to where we update the logical type for clarity.

Maybe if we do `if new_semantic_tags is None` then `new_semantic_tags = _reset_semantic_tags(..)` then we can have the concept of "if we change the logical type the tags get reset" more closely tied together

Comment (thehomebrewnerd) [reply woodwork/schema.py:195]: Yep, sounds good!

Comment (thehomebrewnerd) [reply woodwork/schema.py:195]: As a side note here, related to a prior conversation, I think some of this confusion stems from having a method that is responsible for doing multiple things - if we had this logic completely separated into two methods this wouldn't be quite as confusing.

Comment (tamargrey) [reply woodwork/schema.py:195]: With DataTables, I think we started out with separate methods and combined them to a single `set_types`... maybe @gsheni remembers why??

File: woodwork/utils.py
Comment (thehomebrewnerd) [inline woodwork/utils.py:329]: Old comment?

**Review by thehomebrewnerd — APPROVED** [review]
One minor thing, but otherwise looks good to merge after fixing the merge conflicts.

File: woodwork/utils.py
Comment (tamargrey) [reply woodwork/utils.py]: haha yeppp

**Pull request merged**
